### PR TITLE
feat : Added ability for Gulp workflow to inject constants into app module based on json config files

### PR DESCRIPTION
**Pull request by chalettu**
This type of functionality had been discussed before in previously closed tickets but no one actually submitted a pull req for this functionality.  I had a need to be able to have different config files based on deployment environment and wanted to enhance the existing workflow to support this so I can easily generate constants based on an environment parameter.

Please let me know your thoughts and feedback.  

This would essentially work if you added a --env xxx to whichever gulp command you run from the generator.  
Your generator has a bunch of moving parts, let me know if you think I have missed anything or think it needs some further enhancements.

### CODE REVIEW

**Comment by Swiip** [comment]
It has been discussed many times. For me, it was more for a recipes than a built in feature. But I'm open to discuss it once again.

**Comment by chalettu** [comment]
Thank you for considering it.  I will take a look at updating the tests so they pass.  This feature of environments was pretty important and critical for the angular mobile apps and web sites I write.  With this little addition it makes it easier for me to manage having multiple backends.  I would think this is a fairly common problem.  How to handle different env.  I like using this gulp workflow to handle it because it makes it easy to add this parameter on my CI builds when I deploy to production or any other Env.

File: app/templates/gulp/_env-config.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+var argv = require('yargs').argv;
+var $ = require('gulp-load-plugins')();
+
+//var ngConstant = require('gulp-ng-constant');
+
+var environment = argv.env || 'default';
+//var ENV = JSON.parse(fs.readFileSync('./src/app/env_configs/config-' + environment + '.json', 'utf8')).ENV;
+
+gulp.task('env-config', function () {
+ return gulp.src('src/app/env_configs/config-' + environment + '.json')
```
Comment (loren138) [inline app/templates/gulp/_env-config.js]: This line should read from the config

    return gulp.src(path.join(conf.paths.src, '/app/env_configs/config-' + environment + '.json'))

File: app/templates/gulp/_env-config.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    .pipe($.ngConstant({
+      name: '<%- appName %>',
+      deps: false,
+      dest: 'constants.config.js'
+    }))
+    .pipe(gulp.dest('src/app'));
```
Comment (loren138) [inline app/templates/gulp/_env-config.js]: And this one
.pipe(gulp.dest(path.join(conf.paths.src, '/app')));

File: app/templates/gulp/_env-config.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+'use strict';
+
+var gulp = require('gulp');
+var fs = require('fs');
+var argv = require('yargs').argv;
+var $ = require('gulp-load-plugins')();
```
Comment (loren138) [inline app/templates/gulp/_env-config.js]: To allow reading from config add

   var conf = require('./conf');
   var path = require('path');

**Comment by loren138** [comment]
This is exactly what I needed so I manually merged it in.  That produced a few comments (above) on some code lines.  It would also be nice if it didn't put the config json files in my dist folder since they are already merged.  It would also be nice of the rendered constants.config.js wasn't in the src folder (maybe it could go in .tmp or serve?)  I don't understand gulp enough to fix either of those though.  Overall, thanks so much.  There is no way I could have gotten this to work without this PR.

**Comment by zckrs** [comment]
https://github.com/Swiip/generator-gulp-angular/blob/master/docs/recipes/environment-specific-build.md

A recipe exist now

**Pull request closed without merging**